plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/kvm/ha/KvmHaHelper.java

```diff
@@ -0,0 +1,194 @@
+     * <br><br>
+     * Admins can change the accepted ration of problematic hosts via global settings by updating configuration: "kvm.ha.accepted.problematic.hosts.ratio".
+     */
+    protected boolean isClusteProblematic(Host host) {
+        List<HostVO> hostsInCluster = resourceManager.listAllHostsInCluster(host.getClusterId());
+        List<HostVO> problematicNeighbors = listProblematicHosts(hostsInCluster);
+        int problematicHosts = problematicNeighbors.size();
+        int problematicHostsRatioAccepted = (int) (hostsInCluster.size() * KVMHAConfig.KvmHaAcceptedProblematicHostsRatio.value());
+
+        if (problematicHosts > problematicHostsRatioAccepted) {
+            ClusterVO cluster = clusterDao.findById(host.getClusterId());
+            LOGGER.warn(String.format("%s is problematic but HA will not fence/recover due to its cluster [id: %d, name: %s] containing %d problematic hosts (Down, Disconnected, "
+                            + "Alert or Error states). Maximum problematic hosts accepted for this cluster is %d.",
+                    host, cluster.getId(), cluster.getName(), problematicHosts, problematicHostsRatioAccepted));
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the given Host KVM-HA-Helper is reachable by another host in the same cluster.
+     */
+    protected boolean isHostAgentReachableByNeighbour(Host host) {
+        List<HostVO> neighbors = resourceManager.listHostsInClusterByStatus(host.getClusterId(), Status.Up);
+        for (HostVO neighbor : neighbors) {
+            boolean isVmActivtyOnNeighborHost = isKvmHaAgentHealthy(neighbor);
+            if (isVmActivtyOnNeighborHost) {
+                boolean isReachable = kvmHaAgentClient.isHostReachableByNeighbour(neighbor, host);
+                if (isReachable) {
+                    String.format("%s is reachable by neighbour %s. If CloudStack is failing to reach the respective host then it is probably a network issue between the host "
+                            + "and CloudStack management server.", host, neighbor);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the host is healthy. The health-check is performed via HTTP GET request to a service that retrieves Running KVM instances via Libvirt. <br>
+     * The health-check is executed on the KVM node and verifies the amount of VMs running and if the Libvirt service is running.
+     */
+    public boolean isKvmHealthyCheckViaLibvirt(Host host) {
+        boolean isKvmHaAgentHealthy = isKvmHaAgentHealthy(host);
+
+        if (!isKvmHaAgentHealthy) {
+            if (isClusteProblematic(host) || isHostAgentReachableByNeighbour(host)) {
+                return true;
+            }
+        }
+
+        return isKvmHaAgentHealthy;
```

Comment:
We could use a ternary on return here.

Reply:
Personally, I do prefer _IFs_ than a _ternary_. At least in such cases where a ternary transforms into a huge line. But that might be just me used to the "old fashion" _IFs_ ... :thinking:.

But I have nothing against ternary, we could definitely do such a change.

1. From:
```
public boolean isKvmHealthyCheckViaLibvirt(Host host) {
    boolean isKvmHaAgentHealthy = isKvmHaAgentHealthy(host);

    if (!isKvmHaAgentHealthy) {
        if (isClusteProblematic(host) || isHostAgentReachableByNeighbour(host)) {
        return true;
        }
    }
    
    return isKvmHaAgentHealthy;
}
```

2. To:
```
public boolean isKvmHealthyCheckViaLibvirt(Host host) {
    boolean isKvmHaAgentHealthy = isKvmHaAgentHealthy(host);
    return !isKvmHaAgentHealthy && (isClusteProblematic(host) || isHostAgentReachableByNeighbour(host)) ? true : isKvmHaAgentHealthy;
}
```

OR
```
public boolean isKvmHealthyCheckViaLibvirt(Host host) {
    boolean isKvmHaAgentHealthy = isKvmHaAgentHealthy(host);
    return !isKvmHaAgentHealthy && (isClusteProblematic(host) || isHostAgentReachableByNeighbour(host))
    ? true
    : isKvmHaAgentHealthy;
}
```

Reply:
@GabrielBrascher I see no problem in keeping the `if` statement. In this case, I would only suggest you to unify the `ifs`, to avoid nesting:

```suggestion
        boolean isKvmHaAgentHealthy = isKvmHaAgentHealthy(host);

        if (!isKvmHaAgentHealthy && (isClusteProblematic(host) || isHostAgentReachableByNeighbour(host))) {
            return true;
        }

        return isKvmHaAgentHealthy;
```

From this unify came to me the idea of a ternary...

Reply:
@GutoVeronezi your suggestion makes sense. I've updated the code removing the nested IF.
Thanks for the review!